docs: updates go.md and go-built-in.md as part of issue 9272 (#9688)
tendermint/tendermint: docs/tutorials/go-built-in.md

Applications written in Go can also use Tendermint as a library and run the service in the same 
process as the application. 

By following along this tutorial you will create a Tendermint Core application called kvstore, 
a (very) simple distributed BFT key-value store.
The application will be written in Go and  
some understanding of the Go programming language is expected.
If you have never written Go, you may want to go through [Learn X in Y minutes
Where X=Go](https://learnxinyminutes.com/docs/go/) first, to familiarize
yourself with the syntax.

Note: Please use the latest released version of this guide and of Tendermint.
We strongly advise against using unreleased commits for your development.

### Built-in app vs external app

On the one hand, to get maximum performance you can run your application in 
the same process as the Tendermint Core, as long as your application is written in Go. 
[Cosmos SDK](https://github.com/cosmos/cosmos-sdk) is written
this way. 
This is the approach followed in this tutorial.

On the other hand, having a separate application might give you better security 

> guarantees as two processes would be communicating via established binary protocol. 
> Tendermint Core will not have access to application's state. 
> If that is the way you with to proceed, use the [Writing a built-in Tendermint Core application in Go](./go.md) guide instead of this one.

Reviewer: ```suggestion
guarantees as two processes would be communicating via an established binary protocol. 
Tendermint Core will not have access to application's state. 
If that is the way you wish to proceed, use the [Writing a built-in Tendermint Core application in Go](./go.md) guide instead of this one.
```

> If that is the way you with to proceed, use the [Writing a built-in Tendermint Core application in Go](./go.md) guide instead of this one.

Reviewer: Should this not be external app?

Author: indeed. Updated.

where an instance of the `KVStoreApplication` type is created.

```go
func main() {
    fmt.Println("Hello, Tendermint Core")

    _ = NewKVStoreApplication()
}
```

You can recompile and run the application now by running `go get` and `go build`, but it does
not do anything.
So let's revisit the code adding the logic needed to implement our minimal key/value store
and to start it along with the Tendermint Service.

### 1.3.1 Add a persistent data store
Our application will need to write its state out to persistent storage so that it
can stop and start without losing all of its data.

For this tutorial, we will use [BadgerDB](https://github.com/dgraph-io/badger), a
a fast embedded key-value store. 

First, add Badger as a dependency of your go module using the `go get` command:

`go get github.com/dgraph-io/badger/v3`

Next, let's update the application and its constructor to receive a handle to the database, as follows: 

```go
type KVStoreApplication struct {
	db           *badger.DB
	onGoingBlock *badger.Txn
}

var _ abcitypes.Application = (*KVStoreApplication)(nil)

func NewKVStoreApplication(db *badger.DB) *KVStoreApplication {
	return &KVStoreApplication{db: db}
}
```

> The `onGoingBlock` keeps track of the transaction that will update the application's state when a block 

Author: This is the store badger transaction. I will add the word badger.

> The `onGoingBlock` keeps track of the transaction that will update the application's state when a block 

Author: ```suggestion
The `onGoingBlock` keeps track of the Badger transaction that will update the application's state when a block 
```

> The `onGoingBlock` keeps track of the transaction that will update the application's state when a block 

Author: ```suggestion
The `onGoingBlock` keeps track of the Badger transaction that will update the application's state when a block 
```

is completed. Don't worry about it for now, we'll get to that later.

Next, update the `import` stanza at the top to include the Badger library:

```go
import(
	"github.com/dgraph-io/badger/v3"
	abcitypes "github.com/tendermint/tendermint/abci/types"
)
```

Finally, update the `main.go` file to invoke the updated constructor:

```go
	_ = NewKVStoreApplication(nil)
```

### 1.3.2 CheckTx
When Tendermint Core receives a new transaction from a client, Tendermint asks the application if 

> the transaction is acceptable, using the `CheckTx` method.

Author: ```suggestion
the transaction is acceptable, using the `CheckTx` method.
Invalid transactions will not be shared with other nodes and will not become part of any blocks and, therefore, will not be executed by the application.
```

	code := app.isValid(req.Tx)
	return abcitypes.ResponseCheckTx{Code: code}
}
```

While this `CheckTx` is simple and only validates that the transaction is well-formed, 
it is very common for `CheckTx` to make more complex use of the state of an application.
For example, you may refuse to overwrite an existing value, or you can associate 
versions to the key/value pairs and allow the caller to specify a version to
perform a conditional update.

Depending on the checks and on the conditions violated, the function may return
different values, but any response with a non-zero code will be considered invalid 
by Tendermint. Our `CheckTx` logic returns 0 to Tendermint when a transaction passes 
its validation checks. The specific value of the code is meaningless to Tendermint. 
Non-zero codes are logged by Tendermint so applications can provide more specific 
information on why the transaction was rejected.

Note that `CheckTx` does not execute the transaction, it only verifies that that the transaction could be executed. We do not know yet if the rest of the network has agreed to accept this transaction into a block.

> Finally, make sure to add the bytes package to the your `import` stanza at the top of `app.go`:

Reviewer 2: ```suggestion
Finally, make sure to add the bytes package to the `import` stanza at the top of `app.go`:
```

confuse them with the transactions included in the blocks delivered by Tendermint,
the _application transactions_.

First, let's create a new Badger transaction during `BeginBlock`. All application transactions in the
current block will be executed within this Badger transaction.
Then, return informing Tendermint that the application is ready to receive application transactions:

```go
func (app *KVStoreApplication) BeginBlock(req abcitypes.RequestBeginBlock) abcitypes.ResponseBeginBlock {
	app.onGoingBlock = app.db.NewTransaction(true)
	return abcitypes.ResponseBeginBlock{}
}
```

Next, let's modify `DeliverTx` to add the `key` and `value` to the database transaction every time our application
receives a new application transaction through `RequestDeliverTx`.

```go
func (app *KVStoreApplication) DeliverTx(req abcitypes.RequestDeliverTx) abcitypes.ResponseDeliverTx {
	if code := app.isValid(req.Tx); code != 0 {
		return abcitypes.ResponseDeliverTx{Code: code}
	}

	parts := bytes.SplitN(req.Tx, []byte("="), 2)
	key, value := parts[0], parts[1]

	if err := app.onGoingBlock.Set(key, value); err != nil {

> 		log.Panicf("Error reading database, unable to verify tx: %v", err)

Author: good catch.
```suggestion
		log.Panicf("Error reading database, unable to execute tx: %v", err)
```

execution of the block once accepted.
In the following code, the application simply returns the unmodified group of transactions:

```go
func (app *KVStoreApplication) PrepareProposal(proposal abcitypes.RequestPrepareProposal) abcitypes.ResponsePrepareProposal {
	return abcitypes.ResponsePrepareProposal{Txs: proposal.Txs}
}
```

Once a proposed block is received by a node, the proposal is passed to the application to give
its blessing before voting to accept the proposal.

This mechanism may be used for different reasons, for example to deal with blocks manipulated 
by malicious nodes, in which case the block should not be considered valid.
The following code simply accepts all proposals:

```go
func (app *KVStoreApplication) ProcessProposal(proposal abcitypes.RequestProcessProposal) abcitypes.ResponseProcessProposal {
	return abcitypes.ResponseProcessProposal{Status: abcitypes.ResponseProcessProposal_ACCEPT}
}
```

## 1.4 Starting an application and a Tendermint Core instance in the same process

Now that we have the basic functionality of our application in place, let's put it all together inside of our main.go file.

> Change the contents of your`main.go` file to the following.

Reviewer 2: ```suggestion
Change the contents of your `main.go` file to the following.
```

	"os"
	"os/signal"
	"path/filepath"
	"syscall"

	"github.com/dgraph-io/badger/v3"
	"github.com/spf13/viper"
	cfg "github.com/tendermint/tendermint/config"
	tmflags "github.com/tendermint/tendermint/libs/cli/flags"
	tmlog "github.com/tendermint/tendermint/libs/log"
	nm "github.com/tendermint/tendermint/node"
)

var homeDir string

func init() {
	flag.StringVar(&homeDir, "tm-home", "", "Path to the tendermint config directory (if empty, uses $HOME/.tendermint)")
}

func main() {
	flag.Parse()
	if homeDir == "" {
		homeDir = os.ExpandEnv("$HOME/.tendermint")
	}
	config := cfg.DefaultConfig()

	config.SetRoot(homeDir)

	viper.SetConfigFile(fmt.Sprintf("%s/%s", homeDir, "config/config.toml"))
	if err := viper.ReadInConfig(); err != nil {
		log.Fatalf("Reading config: %v", err)
	}
	if err := viper.Unmarshal(config); err != nil {
		log.Fatalf("Decoding config: %v", err)
	}
	if err := config.ValidateBasic(); err != nil {
		log.Fatalf("Invalid configuration data: %v", err)
	}

	dbPath := filepath.Join(homeDir, "badger")
	db, err := badger.Open(badger.DefaultOptions(dbPath))
	if err != nil {
		log.Fatalf("Opening database: %v", err)
	}
	defer func() {
		if err := db.Close(); err != nil {

> 			log.Fatalf("Closing database: %v", err)

Author: ```suggestion
			log.Printf("Closing database: %v", err)
```
